### PR TITLE
Add test fixtures and an initial e2e test

### PR DESCRIPTION
**Pull request by mike-sul**
- Add test fixtures required for integration/e2e tests;
 - Add an initial e2e test (GCS is faked).

### CODE REVIEW

File: pkg/oshub/upload.go [hunk 1]
```diff
@@ -60,9 +62,19 @@ var (
 	}
 )
 
-func InitUploader(bucket string, workerNumb int) {
+func InitUploader(bucket string, workerNumb int, opts ...StorageFactory) {
```
Comment (doanac) [inline pkg/oshub/upload.go:66]: is the `opts ...` a clever way to have an optional parameter to this function or is there a bigger idea you have in mind here?

File: test/fixtures/ostreehub.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	if err := os.MkdirAll(repoDir, 0755); err != nil {
+		return nil, err
+	}
+
+	oshub.InitHandler(repoDir)
+
+	e := echo.New()
+	e.GET("/v1/repos/:id", oshub.CheckRepoHandler())
+	e.PUT("/v1/repos/:id", oshub.UploadRepoHandler())
+	e.PUT("/v1/repos/:id/summary", oshub.UpdateSummary())
```
Comment (doanac) [inline test/fixtures/ostreehub.go:45]: seems like we might want to create an `oshub.RegisterHandlers` function to make sure these stay in sync with main.go?

File: test/fixtures/storage.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+			if rec := recover(); rec != nil {
+				log.Printf("Failed: %s, %s, %s\n", r.Method, r.URL.Path, rec)
+			}
+		}()
+
+		switch r.Method {
+		case "GET":
+			{
+				s.doGET(w, r)
+			}
```
Comment (doanac) [inline test/fixtures/storage.go:78]: have you run `go fmt` on this? I think this would be formatted differently

**Review by doanac — COMMENTED** [review]
clever. i didn't think about mocking the actual GCS server. 

You think you can add a Dockerfile.e2e or something that has the dependencies needed for running this? Looks like its at least golang + ostree?

**Comment by mike-sul** [comment]
> clever. i didn't think about mocking the actual GCS server.
> 
> You think you can add a Dockerfile.e2e or something that has the dependencies needed for running this? Looks like its at least golang + ostree?

Yes.

File: pkg/oshub/upload.go (hunk 1 above)
Comment (mike-sul) [reply pkg/oshub/upload.go:66]: Just optional param, no any "bigger idea" atm.

File: test/fixtures/storage.go (hunk 3 above)
Comment (mike-sul) [reply test/fixtures/storage.go:78]: I formatted it differently, looks like `go fmt` "fixed it, will double check it as I don't like such formatting, either.

File: main.go [hunk 4]
```diff
@@ -18,19 +17,15 @@ func main() {
 	workerNumb := flag.Int("workers", 20, "A number of concurrent upload workers per a single repo upload request")
 	flag.Parse()
 
-	oshub.InitUploader(*bucket, *workerNumb)
-	oshub.InitHandler(*repoDir)
-	address := net.JoinHostPort("", strconv.Itoa(*port))
-
-	log.Println("address =", address)
+	log.Println("port =", *port)
```
Comment (doanac) [inline main.go:20]: not a big deal. but if you were to move `e := echo.New()` up to here, you could avoid using the stdlib log here.

**Review by doanac — APPROVED** [review]
looks good to me

**Comment by mike-sul** [comment]
> looks good to me

I would like to add another test to test the delta & summary code path.

**Comment by mike-sul** [comment]
@doanac I think that's enough for the first iteration.

File: test/e2e_test.go [hunk 5]
```diff
@@ -52,3 +53,59 @@ func check(t *testing.T, err error, prefix string) {
...
+		"-summary", "-factory", "test-factory", "-server", oshub.URL()).CombinedOutput()
+	if err != nil {
+		t.Errorf("failed to run fiopush: %s, %s\n", err.Error(), string(out))
+	}
+	// TODO: check the generated delta
+	// 1. create a new bare-user repo
+	// 2. pull-local commitN
+	// 3. pull commitN1 from Device Gateway/OSTree Proxy handler
+	// It requires integration of the ostree handler from Device Gateway, we cannot just do pull-local since
+	// the repo layout stored in GCS does not correspond to a layout of an ostree repo stored on a file system
```
Comment (doanac) [inline test/e2e_test.go:110]: you might also just look for the `summary` file to exist. not as complete a test, but gives a little more assurance with not so much effort

**Review by doanac — APPROVED** [review]

File: test/e2e_test.go (hunk 5 above)
Comment (mike-sul) [reply test/e2e_test.go:117]: why not, just updated.

**Pull request merged**